### PR TITLE
Add EventBus resource

### PR DESCRIPTION
**Pull request by embano1**
Description of changes: Add `EventBus` resource.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/resource/event_bus/custom_update_api.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+// resource updates.
+func (rm *resourceManager) customUpdateEventBus(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (*resource, error) {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.customUpdateTrail")
```
Comment (jaypipes) [inline pkg/resource/event_bus/custom_update_api.go:27]:
```suggestion
	exit := rlog.Trace("rm.customUpdateEventBus")
```

File: pkg/resource/event_bus/custom_update_api.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
+package event_bus
```
Comment (jaypipes) [inline pkg/resource/event_bus/custom_update_api.go:1]: I actually think it's best to just put all the custom code into `hooks.go`...

File: pkg/resource/event_bus/custom_update_api.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
+	ackutil "github.com/aws-controllers-k8s/runtime/pkg/util"
+	svcsdk "github.com/aws/aws-sdk-go/service/eventbridge"
+
+	svcapitypes "github.com/aws-controllers-k8s/eventbridge-controller/apis/v1alpha1"
+)
+
+// customUpdateTrail implements a custom logic for handling Trail
+// resource updates.
+func (rm *resourceManager) customUpdateEventBus(
```
Comment (jaypipes) [inline pkg/resource/event_bus/custom_update_api.go:19]: could shorten this to just `customUpdate`, since the package is managing Event Bus resources, it's redundant to suffix things with `EventBus`

File: pkg/resource/event_bus/custom_update_api.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (*resource, error) {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.customUpdateTrail")
+	defer func(err error) { exit(err) }(err)
```
Comment (jaypipes) [inline pkg/resource/event_bus/custom_update_api.go:28]: technically, you could just do this instead:
```suggestion
	defer func() { exit(err) }()
```

File: pkg/resource/event_bus/custom_update_api.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+		err = rm.syncEventBusTags(ctx, latest, desired)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return desired, nil
+}
+
+// syncEventBusTags updates event bus tags
+func (rm *resourceManager) syncEventBusTags(
```
Comment (jaypipes) [inline pkg/resource/event_bus/custom_update_api.go:45]: Maybe just `syncTags` instead of `syncEventBusTags`?

File: pkg/resource/event_bus/custom_update_api.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+	if len(removed) > 0 {
+		_, err = rm.sdkapi.UntagResourceWithContext(
+			ctx,
+			&svcsdk.UntagResourceInput{
+				ResourceARN: (*string)(latest.ko.Status.ACKResourceMetadata.ARN),
+				TagKeys:     sdkTagStringsFromResourceTags(removed),
+			})
+
+		rm.metrics.RecordAPICall("UPDATE", "RemoveTags", err)
```
Comment (jaypipes) [inline pkg/resource/event_bus/custom_update_api.go:64]:
```suggestion
		rm.metrics.RecordAPICall("UPDATE", "UntagResource", err)
```

File: pkg/resource/event_bus/custom_update_api.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+	if len(added) > 0 {
+		_, err = rm.sdkapi.TagResourceWithContext(
+			ctx,
+			&svcsdk.TagResourceInput{
+				ResourceARN: (*string)(latest.ko.Status.ACKResourceMetadata.ARN),
+				Tags:        sdkTagsFromResourceTags(added),
+			})
+
+		rm.metrics.RecordAPICall("UPDATE", "AddTags", err)
```
Comment (jaypipes) [inline pkg/resource/event_bus/custom_update_api.go:78]:
```suggestion
		rm.metrics.RecordAPICall("UPDATE", "TagResource", err)
```

File: pkg/resource/event_bus/custom_update_api.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	exit := rlog.Trace("rm.syncEventBusTags")
+	defer func(err error) { exit(err) }(err)
+
+	added, removed := computeTagsDelta(latest.ko.Spec.Tags, desired.ko.Spec.Tags)
+
+	if len(removed) > 0 {
+		_, err = rm.sdkapi.UntagResourceWithContext(
+			ctx,
+			&svcsdk.UntagResourceInput{
+				ResourceARN: (*string)(latest.ko.Status.ACKResourceMetadata.ARN),
```
Comment (jaypipes) [inline pkg/resource/event_bus/custom_update_api.go:60]: Might be good to pull this into a variable in the function scope:

```go
arn := (*string)(latest.ko.Status.ACKResourceMetadata.ARN)
```

and then simple do this here:

```suggestion
				ResourceARN: arn,
```

File: pkg/resource/event_bus/custom_update_api.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(added) > 0 {
+		_, err = rm.sdkapi.TagResourceWithContext(
+			ctx,
+			&svcsdk.TagResourceInput{
+				ResourceARN: (*string)(latest.ko.Status.ACKResourceMetadata.ARN),
```
Comment (jaypipes) [inline pkg/resource/event_bus/custom_update_api.go:74]: And if you moved the variable into the function scope above...
```suggestion
				ResourceARN: arn,
```

File: pkg/resource/event_bus/delta.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	if ackcompare.HasNilDifference(a.ko.Spec.Name, b.ko.Spec.Name) {
+		delta.Add("Spec.Name", a.ko.Spec.Name, b.ko.Spec.Name)
+	} else if a.ko.Spec.Name != nil && b.ko.Spec.Name != nil {
+		if *a.ko.Spec.Name != *b.ko.Spec.Name {
+			delta.Add("Spec.Name", a.ko.Spec.Name, b.ko.Spec.Name)
+		}
+	}
+	if !reflect.DeepEqual(a.ko.Spec.Tags, b.ko.Spec.Tags) {
+		delta.Add("Spec.Tags", a.ko.Spec.Tags, b.ko.Spec.Tags)
+	}
```
Comment (jaypipes) [inline pkg/resource/event_bus/delta.go:60]: I believe you will need to ignore the `Tags` field for the purpose of comparisons and put in a custom comparator function since `Tags` is a `[]struct{Key,Value}` and `reflect.DeepEqual()` doesn't know how to sort that and therefore differences in sort order (and only sort order) result in an erroneous Delta...

See how the IAM controller handles this [here](https://github.com/aws-controllers-k8s/iam-controller/blob/2f0a8cb2c303d0d5ec8d5133b4b350f20d4752be/generator.yaml#L80-L81) and [here](https://github.com/aws-controllers-k8s/iam-controller/blob/2f0a8cb2c303d0d5ec8d5133b4b350f20d4752be/generator.yaml#L119-L121)

**Review by jaypipes — CHANGES_REQUESTED** [review]
Excellent first resource manager PR, @embano1! :) A few suggestions for you inline. Hit me up on Slack if you have any questions

File: pkg/resource/event_bus/custom_update_api.go (hunk 2 above)
Comment (embano1) [reply pkg/resource/event_bus/custom_update_api.go:1]: Done

File: pkg/resource/event_bus/custom_update_api.go (hunk 3 above)
Comment (embano1) [reply pkg/resource/event_bus/custom_update_api.go:19]: Done. Left extension here though: `exit := rlog.Trace("rm.customUpdateEventBus")`

File: pkg/resource/event_bus/custom_update_api.go (hunk 4 above)
Comment (embano1) [reply pkg/resource/event_bus/custom_update_api.go:28]: Done, copied verbatim back then from `Trail`.

File: pkg/resource/event_bus/custom_update_api.go (hunk 5 above)
Comment (embano1) [reply pkg/resource/event_bus/custom_update_api.go:45]: Done. Kept `exit := rlog.Trace("rm.syncEventBusTags")` though to avoid log parsing ambiguity?

File: pkg/resource/event_bus/custom_update_api.go (hunk 6 above)
Comment (embano1) [reply pkg/resource/event_bus/custom_update_api.go:64]: Done.

File: pkg/resource/event_bus/custom_update_api.go (hunk 7 above)
Comment (embano1) [reply pkg/resource/event_bus/custom_update_api.go:78]: Done.

File: pkg/resource/event_bus/custom_update_api.go (hunk 8 above)
Comment (embano1) [reply pkg/resource/event_bus/custom_update_api.go:60]: Done.

File: pkg/resource/event_bus/custom_update_api.go (hunk 9 above)
Comment (embano1) [reply pkg/resource/event_bus/custom_update_api.go:74]: Yup, done thx to Goland's nice refactoring features :D

File: pkg/resource/event_bus/delta.go (hunk 10 above)
Comment (embano1) [reply pkg/resource/event_bus/delta.go:60]: Done

**Review by jaypipes — APPROVED** [review]
Wunderbar! :) :+1:

**Review by A-Hilaly — APPROVED** [review]
